lib/internal/process/execution.js

```diff
@@ -238,10 +246,120 @@ function readStdin(callback) {
+ * @param {Error} originalStack the stack to decorate
+ * @param {string} newMessage the message to add to the error stack
+ * @returns {void}
+ */
+function decorateCJSErrorWithTSMessage(originalStack, newMessage) {
+  const lines = StringPrototypeSplit(originalStack, '\n');
+  ArrayPrototypeSplice(lines, 3, 0, newMessage);
+  return ArrayPrototypeJoin(lines, '\n');
+}
+
+/**
+ *
+ * Wrapper of evalScript
+ *
+ * This function wraps the evaluation of the source code in a try-catch block.
+ * If the source code fails to be evaluated, it will retry evaluating the source code
+ * with the TypeScript parser.
+ *
+ * If the source code fails to be evaluated with the TypeScript parser,
+ * it will rethrow the original error, adding the TypeScript error message to the stack.
+ *
+ * This way we don't change the behavior of the code, but we provide a better error message
+ * in case of a typescript error.
+ */
+function evalTypeScript(name, source, breakFirstLine, print, shouldLoadESM = false) {
+  try {
+    evalScript(name, source, breakFirstLine, print, shouldLoadESM);
+  } catch (originalError) {
+    try {
+      const strippedSource = stripTypeScriptModuleTypes(source, name, false);
+      evalScript(name, strippedSource, breakFirstLine, print, shouldLoadESM);
+      // Emit the experimental warning after the code was successfully evaluated.
+      emitExperimentalWarning('Type Stripping');
+    } catch (tsError) {
+      if (tsError.code === 'ERR_INVALID_TYPESCRIPT_SYNTAX') {
+        originalError.stack = decorateCJSErrorWithTSMessage(originalError.stack, tsError.message);
+      }
+      throw originalError;
+    }
+  }
+}
+
+/**
+ * Wrapper of evalModuleEntryPoint
+ *
+ * This function wraps the evaluation of the source code in a try-catch block.
+ * If the source code fails to be evaluated, it will retry evaluating the source code
+ * with the TypeScript parser.
+ *
+ */
+function evalTypeScriptModuleEntryPoint(source, print) {
+  if (print) {
+    throw new ERR_EVAL_ESM_CANNOT_PRINT();
+  }
+
+  RegExpPrototypeExec(/^/, ''); // Necessary to reset RegExp statics before user code runs.
+
+  return require('internal/modules/run_main').runEntryPointWithESMLoader(
+    async (loader) => {
+      try {
+        // Await here to catch the error and rethrow it with the typescript error message.
+        return await loader.eval(source, getEvalModuleUrl(), true);
+      } catch (originalError) {
+        try {
+          const url = getEvalModuleUrl();
+          const strippedSource = stripTypeScriptModuleTypes(source, url, false);
+          const result = await loader.eval(strippedSource, url, true);
+          // Emit the experimental warning after the code was successfully evaluated.
+          emitExperimentalWarning('Type Stripping');
+          return result;
+        } catch (tsError) {
+          if (tsError.code === 'ERR_INVALID_TYPESCRIPT_SYNTAX') {
+            originalError.stack = `${tsError.message}\n\n${originalError.stack}`;
```

Comment:
Same here, we should not assume the caught value will be an `Error` object

Reply:
I wonder what the right behavior should be when re-throwing the original error